File: data/C++/Layer.cpp
#include "Layer.h"
#include <regex>
#include "PAMFormatter.h"
#include "XMLFormatter.h"
#include "BMPFormatter.h"
#include "Exceptions.h"

//int Layer::maxHeight = 0;
//int Layer::maxWidth = 0;
//int Layer::numOfLayers = 0;
int Layer::ID = 0;

Layer::~Layer() {

	auto pom = image.second;
	delete pom;
}

Layer::Layer(int height, int width) {

	transparency = 1;	//not transparent layer if no starting image
	image = std::make_pair(HeightWidth(height, width), new std::list<Pixel>);

	Pixel transp(255, 255, 255,255);

	for (int i = 0; i < image.first.height * image.first.width; i++) {
		image.second->push_back(transp);
	}

	//maxHeight = (height > maxHeight) ? height : maxHeight;
	//maxWidth = (width > maxWidth) ? width : maxWidth;

	//increaseSize(height,width);
}


Layer::Layer(std::string source) {

	std::regex rx("[^.]*.([a-z]*)");
	std::smatch result;

	std::regex_match(source, result, rx);
	std::string format = result.str(1);

	Formatter* formatter = nullptr;

	if (format == "pam")
		formatter = new PAMFormatter();
	else if (format == "bmp")
		formatter = new BMPFormatter();
	else if (format == "xml")
		formatter = new XMLFormatter();
	else
		throw NotAppropriateFormat();

	image = formatter->read(source);

	//maxHeight = (height > maxHeight) ? height : maxHeight;
	//maxWidth = (width > maxWidth) ? width : maxWidth;

	//increaseSize();
}

void Layer::increaseSize(int maxHeight, int maxWidth) {

	increaseWidth(maxHeight, maxWidth);
	increaseHeight(maxHeight, maxWidth);
}

void Layer::increaseWidth(int maxHeight, int maxWidth) {

	if (image.first.width >= maxWidth)
		return;

	int diff = maxWidth - image.first.width;

	std::list<Pixel> padding;

	for (int i = 0; i < diff; i++) {
		padding.push_back(Pixel(255, 255, 255, 0));	//fully transparent pixel
	}


	for (int i = 0; i < image.first.height - 1; i++) {

		int offset = i * diff + (i + 1) * image.first.width;
		auto it = image.second->begin();
		std::advance(it, offset);

		image.second->insert(it, padding.begin(), padding.end());
	}

	for (auto pom : padding)
		image.second->push_back(pom);

	image.first.width = maxWidth;
}

void Layer::increaseHeight(int maxHeight, int maxWidth) {

	if (image.first.height >= maxHeight)
		return;

	int diff = maxHeight - image.first.height;

	for (int i = 0; i < diff * maxWidth; i++) {

		image.second->push_back(Pixel(255, 255, 255, 0));	//fully transparent pixel
	}

	image.first.height = maxHeight;
}

File: obradaSlike/Layer.java
package obradaSlike;

import java.awt.image.BufferedImage;

public class Layer {
	
	private BufferedImage image;
	private boolean visible = true;
	private boolean active = true;
	private double transparency = 1;	//range from 0-1(1 means it's not transparent)
	private static int ID = 0;
	private int id = ID++;
	
	public boolean isActive() { return active; }
	
	public double getTransparency() { return transparency; }
	
	public boolean isVisible(){ return visible; }
	
	public int getWidth(){ return image.getWidth(); }
	
	public int getHeight() { return image.getHeight(); }

	public int getId() { return id; }
	
	public void setActive(boolean val) { active = val; }
	
	public void setVisible(boolean val) { visible = val; }
	
	public void setTransparency(double tr) { transparency = (tr > 1 || tr < 0) ? 1 : tr; }
	
	public BufferedImage getImage() {
		return image;
	}
	
	public Layer(String source) throws ErrorOccurred {
		String format = source.substring(source.length() - 3);
		Formatter formatter = null;
		
		if (format.equals("bmp"))
			formatter = new BMPFormatter();
		else if (format.equals("pam"))
			formatter = new PAMFormatter();
		else if (format.equals("xml"))
			formatter = new XMLFormatter();
		else
			throw new ErrorOccurred("File format not recognized");
		
		image = formatter.read(source);
	}
	
	public Layer(int height,int width) {
		image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
		int val = (0xff << 24) | (0xff << 16) | (0xff << 8) | (0xff);
		
		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width ; j++) {
				image.setRGB(j, i, val);
			}
		}
	}
	
	public void increaseSize(int maxWidth,int maxHeight) {
		if (maxWidth == image.getWidth() && maxHeight == image.getHeight())
			return;
		
		BufferedImage newImage = new BufferedImage(maxWidth,maxHeight,BufferedImage.TYPE_INT_ARGB);
		
		for (int i = 0; i < image.getHeight(); i++) {
			for (int j = 0; j < image.getWidth(); j++) {
				newImage.setRGB(j, i, image.getRGB(j, i));
			}
		}
		
		int transpPixel = (0xff << 16) | (0xff << 8) | (0xff);
		
		for (int i = image.getHeight() ; i < newImage.getHeight() ; i++) {
			for (int j = image.getWidth() ; j < newImage.getWidth() ; j++) {
				newImage.setRGB(j, i, transpPixel);
			}
		}
		
		image = newImage;
	}
	
}
